File: client/src/main/java/object/AchievementBadge.java
package object;

import util.AchievementsUtil;
import util.Registry;
import utils.Achievement;

import javax.swing.*;
import java.awt.*;

import static util.AchievementUtilKt.LOCKED_ICON;

public class AchievementBadge extends JLabel
							  implements Registry
{
	private final String registryLocation;
	private final String name;
	private final String explanation;
	private final String description;
	
	public AchievementBadge(Achievement achievement)
	{
		setPreferredSize(new Dimension(56, 56));
		
		this.registryLocation = achievement.getRegistryLocation();
		this.name = achievement.getTitle();
		this.explanation = achievement.getExplanation();
		this.description = achievement.getDescription();

		if (achievements.getBoolean(registryLocation, false)) {
			setIcon(AchievementsUtil.getIconForAchievement(registryLocation));
		} else {
			setIcon(LOCKED_ICON);
		}
	}
	
	@Override
	public String toString() 
	{
		return name;
	}
	
	public boolean isEarned() 
	{
		return achievements.getBoolean(registryLocation, false);
	}
	@Override
	public String getName() 
	{
		return name;
	}
	public String getExplanation()
	{
		return explanation;
	}
	public String getDescription()
	{
		return description;
	}
}

File: client/src/main/kotlin/bean/AchievementBadge.kt
package bean

import achievement.getIcon
import achievement.isUnlocked
import java.awt.Dimension
import javax.swing.JLabel
import util.Images
import utils.Achievement

class AchievementBadge(private val achievement: Achievement) : JLabel() {
    val explanation = achievement.explanation
    val description = achievement.description

    init {
        preferredSize = Dimension(56, 56)

        this.name = achievement.title

        toggle()
    }

    fun toggle() {
        icon = if (isEarned) achievement.getIcon() else Images.ACHIEVEMENT_LOCKED

        if (achievement.hidden) {
            isVisible = isEarned
        }

        repaint()
    }

    val isEarned: Boolean
        get() = achievement.isUnlocked()
}
